README.md
prometheus-webhook-snmp is a Prometheus Alertmanager receiver that translates incoming notifications into SNMP v2c traps.

# Features

- Independent from specific SNMP OID's
- The SNMP OID's of variable binds can be customized via prefix
- Extract SNMP OID from alert label
- Use a default SNMP OID for all alerts without an OID label
- Export Prometheus metrics from this receiver

# Start the receiver

To start the receiver execute the command:

    $ ./prometheus-webhook-snmp run

To start with HTTPS, pass the certificate and key parameters:

    $ ./prometheus-webhook-snmp --cert "certificate.pem" --key "key.pem" run

Note: Both '--cert' and '--key' are needed to enable HTTPS.
# Send a test SNMP trap

If you want to send a test SNMP trap, then simply execute the following command. This can be used to test your command line parameters.

    $ ./prometheus-webhook-snmp test

# Command line parameters
Command line parameters have precedence over global configuration settings.

## Global

### --snmp-host
The host (IPv4, IPv6 or FQDN) of the SNMP trap receiver. Defaults to ``localhost``.

### --snmp-port
The port of the SNMP trap receiver. Defaults to ``162``.

### --snmp-community
The SNMP community string. Defaults to ``public``.

### --snmp-retries
Maximum number of request retries. Defaults to ``5``.

### --snmp-timeout
Response timeout in seconds. Defaults to ``1``.

### --alert-oid-label
The label where to find the OID. Defaults to ``oid``.

Example Prometheus alert rule:

```yaml
groups:
- name: mon
  rules:
  - alert: low monitor quorum count
    expr: sum(ceph_mon_quorum_status) < 3
    labels:
      severity: critical
      type: ceph_default
      oid: 1.3.6.1.4.1.50495.15.1.2.3.1
    annotations:
      description: Monitor count in quorum is low.
```

### --trap-oid-prefix
The OID prefix for trap variable bindings. Defaults to ``1.3.6.1.4.1.50495.15``.

### --trap-default-oid
The trap OID if none is found in the Prometheus alert labels. Defaults to ``1.3.6.1.4.1.50495.15.1.2.1``.

### --trap-default-severity
The trap severity if none is found in the Prometheus alert labels. Defaults to an empty string.

## Command ``run``

### --host
The host to use, defaults to ``localhost``.

### --port
Port to listen for Prometheus Alertmanager notifications. Defaults to ``9099``.

Example Prometheus Alertmanager configuration:

```yaml
receivers:
  - name: 'prometheus-webhook-snmp'
    webhook_configs:
    - url: 'http://localhost:9099'
```

### --metrics
Provide Prometheus metrics from this receiver under the URL ``/metrics``.

Example Prometheus configuration:

```yaml
scrape_configs:
  - job_name: 'prometheus-webhook-snmp'
    scrape_interval: 30s
    static_configs:
      - targets: ['localhost:9099']
```

# Global configuration file
The Prometheus Alertmanager receiver can be configured via configuration file, too.
Command line arguments have precedence over the settings in the configuration file.
The configuration file is written in YAML format. The file will be loaded in the
following order and precedence from ``/etc/prometheus-webhook-snmp.conf`` and the
directory in which the prometheus-webhook-snmp command is located.
Parameters in these files have precedence over default configuration settings.
Please replace hyphens in parameter names with underscores.

Example configuration:

```yaml
debug: True
snmp_retries: 1
snmp_community: private
host: promalertmgr.foo.com
port: 9101
```

# SNMP schema

## Traps

| OID | Description |
| :--- | :--- |
| ``trap-oid-prefix``.1.2.1 | The default trap. This is used if no OID is specified in the alert labels. |
| ``trap-oid-prefix``.1.2.[2...N] | Custom traps. |

## Objects

The following objects are appended as variable binds to a SNMP trap.

| OID | Type | Description |
| :--- | :---: | :--- |
| ``trap-oid-prefix``.1.1.1 | String | The name of the Prometheus alert. |
| ``trap-oid-prefix``.1.1.2 | String | The status of the Prometheus alert. |
| ``trap-oid-prefix``.1.1.3 | String | The severity of the Prometheus alert. |
| ``trap-oid-prefix``.1.1.4 | String | Unique identifier for the Prometheus instance. |
| ``trap-oid-prefix``.1.1.5 | String | The name of the Prometheus job. |
| ``trap-oid-prefix``.1.1.6 | String | The Prometheus alert description field. |
| ``trap-oid-prefix``.1.1.7 | String | Additional Prometheus alert labels as JSON string. |
| ``trap-oid-prefix``.1.1.8 | Unix timestamp | The time when the Prometheus alert occurred. |
| ``trap-oid-prefix``.1.1.9 | String | The raw Prometheus alert as JSON string. |


# Docker

## Pull the container

To get the container image, execute the following command:

    $ docker pull registry.opensuse.org/opensuse/prometheus-webhook-snmp:latest
    $ docker pull registry.opensuse.org/opensuse/prometheus-webhook-snmp:1.4

## Networking

You need to publish the port the Prometheus receiver is listening on by
using the command line argument `--publish <HOST_PORT>:<CONTAINER_PORT>`
when running the container because the port is not exposed automatically
because it is configurable. The default port is `9099`.

Alternatively simply connect the container to the host network by using
the command line argument `--network=host`. Additionally the SNMP host
needs to be configured. Use the container's network gateway to be able
to receive SNMP traps outside the container.

## Run the container

There are two options for configuring the container:

1. Environment variables
2. Configuration file

### Environment variables

Use `ARGS` to set [global options](#global) and `RUN_ARGS` for the
[run command options](#command-run). The environment variable `ARGS`
defaults to `--debug`.

    $ docker run --env ARGS="--debug --snmp-community=foo" --env RUN_ARGS="--metrics" ...

### Configuration file

Use a [configuration file](#global-configuration-file) by mounting it into the container.

    $ docker run -v $(pwd)/prometheus-webhook-snmp.conf:/etc/prometheus-webhook-snmp.conf ...


CHANGELOG.md
v1.5
  * Issue #26: Fix 'TypeError' when using '--port' argument.
  * Add HTTPS support. Use the arguments '--cert' and '--key' to configure
    the SSL certificate (in PEM format).

v1.4
  * Issue #22: Prevent Python type errors when sending an SNMP trap.

v1.3

  * Add 'trap-default-severity' configuration option.
  * Harden the notification parser.

v1.2

  * Show used configuration settings in debug mode.
  * Load configuration file from the directory in which the prometheus-webhook-snmp
    command is located.
  * Automatically convert hyphens to underscores in configuration file parameters.

v1.1

  * Add support for global configuration file /etc/prometheus-webhook-snmp.conf
  * Add ability to configure systemd service via /etc/default/prometheus-webhook-snmp
  * Added python3-PyYAML as new dependency
  * python3-pysnmp needs to be at least 4.4.1

v1.0

  * Initial Debian packaging.


tests/test_misc.py
import unittest

import mock
from pyfakefs import fake_filesystem

from prometheus_webhook_snmp.utils import (Config, parse_notification,
                                           send_snmp_trap)

NOTIFICATION_FIRING = {
    'receiver': 'test-01',
    'status': 'firing',
    'alerts': [{
        'status': 'firing',
        'labels': {
            'alertname': 'load_1',
            'job': 'node-exporter',
            'severity': 'warning',
            'oid': '1.3.6.1.4.1.50495.15.1.2.2'
        },
        'annotations': {
            'description': 'aaa'
        },
        'startsAt': '2019-04-02T17:29:05.244307923Z',
        'endsAt': '0001-01-01T00:00:00Z',
        'generatorURL': 'http://arrakis:9090/graph?g0.expr=node_load1+%3E+0&g0.tab=1'
    }],
    'groupLabels': {},
    'commonLabels': {
        'instance': 'localhost:9100',
        'job': 'node-exporter',
        'severity': 'warning'
    },
    'commonAnnotations': {},
    'externalURL': 'http://arrakis:9093',
    'version': '4',
    'groupKey': '{}:{}'
}
NOTIFICATION_RESOLVED = {  # pylint: disable=line-too-long
    'receiver': 'test-02',
    'status': 'resolved',
    'alerts': [{
        'status': 'resolved',
        'labels': {
            'alertname': 'load_2',
            'instance': 'localhost:9100',
            'job': 'node-exporter',
            'severity': 'critical',
            'foo': 'abc',
            'bar': 123
        },
        'annotations': {
            'description': 'aaa',
            'summary': 'bbb'
        },
        'startsAt': '2019-04-01T09:17:46.646300974Z',
        'endsAt': '2019-04-01T09:19:47.126304527Z',
        'generatorURL': 'http://arrakis:9090/graph?g0.expr=node_load1+%3E+2+'
                        'and+node_load1+%3E+4&g0.tab=1'
    }],
    'groupLabels': {},
    'commonLabels': {
        'instance': 'localhost:9100',
        'job': 'node-exporter',
        'severity': 'warning'
    },
    'commonAnnotations': {},
    'externalURL': 'http://arrakis:9093',
    'version': '4',
    'groupKey': '{}:{}'
}


class FunctionTestCase(unittest.TestCase):
    def test_parse_notification_none(self):
        trap_data = parse_notification({}, {
            'alertsss': {}
        })
        self.assertIsNone(trap_data)

    def test_parse_notification(self):
        trap_data = parse_notification({
            'alert_oid_label': 'oid',
            'trap_default_oid': '1.2.3.4',
            'trap_default_severity': ''
        }, NOTIFICATION_FIRING)
        self.assertIsInstance(trap_data, list)
        trap_data = trap_data[0]
        self.assertEqual(trap_data['oid'], '1.3.6.1.4.1.50495.15.1.2.2')
        self.assertEqual(trap_data['alertname'], 'load_1')
        self.assertEqual(trap_data['status'], 'firing')
        self.assertEqual(trap_data['severity'], 'warning')
        self.assertIsNone(trap_data['instance'])
        self.assertEqual(trap_data['job'], 'node-exporter')
        self.assertEqual(trap_data['description'], 'aaa')
        self.assertIsInstance(trap_data['labels'], dict)
        self.assertEqual(trap_data['labels'], {})
        self.assertEqual(trap_data['timestamp'], 1554226145)
        self.assertIsInstance(trap_data['rawdata'], dict)

    def test_parse_notification_no_oid(self):
        trap_data = parse_notification({
            'alert_oid_label': 'oid',
            'trap_default_oid': '1.2.3.4',
            'trap_default_severity': ''
        }, NOTIFICATION_RESOLVED)
        self.assertIsInstance(trap_data, list)
        trap_data = trap_data[0]
        self.assertEqual(trap_data['oid'], '1.2.3.4')
        self.assertEqual(trap_data['alertname'], 'load_2')
        self.assertEqual(trap_data['status'], 'resolved')
        self.assertEqual(trap_data['severity'], 'critical')
        self.assertEqual(trap_data['instance'], 'localhost:9100')
        self.assertEqual(trap_data['job'], 'node-exporter')
        self.assertEqual(trap_data['description'], 'bbb')
        self.assertIsInstance(trap_data['labels'], dict)
        self.assertEqual(trap_data['labels'], {'foo': 'abc', 'bar': 123})
        self.assertEqual(trap_data['timestamp'], 1554110387)
        self.assertIsInstance(trap_data['rawdata'], dict)

    def test_send_snmp_trap(self):
        config = Config()
        self.assertIsInstance(config['trap_default_severity'], str)
        send_snmp_trap(config, {
            'oid': '1.3.6.1.4.1.50495.15.1.2.1',
            'alertname': None,
            'status': 'resolved',
            'severity': config['trap_default_severity'],
            'instance': None,
            'job': None,
            'description': None,
            'labels': {},
            'timestamp': 1554110387,
            'rawdata': {}
        })


class ConfigTestCase(unittest.TestCase):
    fs = fake_filesystem.FakeFilesystem()
    f_open = fake_filesystem.FakeFileOpen(fs)

    def test_defaults(self):
        self.assertIsInstance(Config.defaults(), dict)

    def test_reset(self):
        config = Config()
        config['snmp_community'] = 'private'
        config.reset('snmp_community')
        self.assertEqual(config['snmp_community'],
                         Config.defaults()['snmp_community'])

    def test_reset_all(self):
        config = Config()
        config['foo'] = 'bar'
        config.reset()
        self.assertDictEqual(config, Config.defaults())

    @mock.patch('builtins.open', new=f_open)
    def test_load(self):
        self.fs.create_file('/etc/abc.conf', contents='''foo: bar\n''')
        config = Config()
        config.load('abc')
        self.assertEqual(config['foo'], 'bar')


tests/test_server.py
import json

import cherrypy
from cherrypy.test import helper

from prometheus_webhook_snmp.utils import Context, Root, get_http_server_config


class ServerTestCase(helper.CPWebCase):
    @staticmethod
    def setup_server():
        ctx = Context()
        config = get_http_server_config()
        cherrypy.config.update({'environment': 'test_suite'})
        ServerTestCase.app = cherrypy.tree.mount(Root(ctx), config=config)

    def _request(self, url, method, data=None):
        if not data:
            body = None
            headers = None
        else:
            body = json.dumps(data)
            headers = [('Content-Type', 'application/json'),
                       ('Content-Length', str(len(body)))]
        self.getPage(url, method=method, body=body, headers=headers)

    def _post(self, url, data=None):
        self._request(url, 'POST', data)

    def _get(self, url):
        self._request(url, 'GET')

    def test_post(self):
        self._post('/', data={'alerts': {}})
        self.assertStatus('200 OK')

    def test_get_telemetry_on(self):
        cherrypy.tree.apps[''].root.ctx.config['metrics'] = True
        self._get('/metrics')
        self.assertStatus('200 OK')

    def test_get_telemetry_off(self):
        cherrypy.tree.apps[''].root.ctx.config['metrics'] = False
        self._get('/metrics')
        self.assertStatus('405 Method Not Allowed')


prometheus_webhook_snmp/utils.py
import datetime
import ipaddress
import json
import logging
import os
import sys

import cherrypy
import dateutil.parser
import prometheus_client
import yaml
from pysnmp import hlapi

logger = logging.getLogger(__name__)


def parse_notification(config, notification):
    """
    Parse the Alertmanager notification.

    Example:
    {
      "receiver": "storage",
      "status": "firing",
      "alerts": [
        {
          "status": "firing",
          "labels": {
            "alertname": "OSD down",
            "monitor": "ses",
            "oid": "1.3.6.1.4.1.50495.15.1.2.4.2",
            "severity": "warning",
            "type": "ceph_default"
          },
          "annotations": {
            "description": "One or more OSDs down for more than 15 minutes."
          },
          "startsAt": "2020-03-24T10:25:29.689078708-04:00",
          "endsAt": "0001-01-01T00:00:00Z",
          "generatorURL": "http://xxxx:9090/graph?g0.expr=..."
        }
      ],
      ...
    }

    See https://prometheus.io/docs/alerting/configuration/#webhook_config

    :param config: The configuration data.
    :type config: dict
    :param notification:
    :type notification: dict
    :return: Returns a list of alert objects.
    :rtype: list or None
    """
    result = None
    if notification and 'alerts' in notification:
        result = []
        for alert in notification['alerts']:
            starts_at = dateutil.parser.parse(alert['startsAt'])
            if alert['endsAt'] != '0001-01-01T00:00:00Z':
                ends_at = dateutil.parser.parse(alert['endsAt'])
            else:
                ends_at = None

            if alert['status'] == 'firing':
                time = starts_at
            elif alert['status'] == 'resolved':
                time = ends_at
            else:
                time = ends_at or starts_at

            labels = alert['labels']
            annotations = alert['annotations']

            alertname = labels.pop('alertname', None)
            summary = annotations.pop('summary', None)
            description = annotations.pop('description', None)
            timestamp = int(time.replace(
                tzinfo=datetime.timezone.utc).timestamp())

            result.append({
                'oid': labels.pop(config['alert_oid_label'],
                                  config['trap_default_oid']),
                'alertname': alertname,
                'status': alert['status'],
                'severity': labels.pop('severity',
                                       config['trap_default_severity']),
                'instance': labels.pop('instance', None),
                'job': labels.pop('job', None),
                'description': summary or description,
                'labels': labels,
                'timestamp': timestamp,
                'rawdata': alert
            })
    else:
        # This should never happen. Don't raise an exception because it does not
        # make sense to notify Prometheus Alertmanager about this.
        # raise cherrypy.HTTPError(400, 'No alerts in Prometheus notification payload')
        logger.error('No alerts in Prometheus notification payload: %s',
                     json.dumps(notification))
    return result


def send_snmp_trap(config, trap_data):
    """
    Send a SNMP v2c trap.
    :param config: The configuration data.
    :type config: dict
    :param trap_data:
    """
    oids = {
        'alertname': '{}.1.1.1'.format(config['trap_oid_prefix']),
        'status': '{}.1.1.2'.format(config['trap_oid_prefix']),
        'severity': '{}.1.1.3'.format(config['trap_oid_prefix']),
        'instance': '{}.1.1.4'.format(config['trap_oid_prefix']),
        'job': '{}.1.1.5'.format(config['trap_oid_prefix']),
        'description': '{}.1.1.6'.format(config['trap_oid_prefix']),
        'labels': '{}.1.1.7'.format(config['trap_oid_prefix']),
        'timestamp': '{}.1.1.8'.format(config['trap_oid_prefix']),
        'rawdata': '{}.1.1.9'.format(config['trap_oid_prefix'])
    }

    transport_addr = (config['snmp_host'], config['snmp_port'])

    # Configure the transport target (IPv4 or IPv6).
    try:
        # Will raise an exception if ``snmp_host`` isn't an IPv6 address.
        ipaddress.IPv6Address(config['snmp_host'])
        transport_target = hlapi.Udp6TransportTarget(transport_addr)
    except ValueError:
        transport_target = hlapi.UdpTransportTarget(transport_addr)
    transport_target.retries = config['snmp_retries']
    transport_target.timeout = config['snmp_timeout']

    var_binds = hlapi.NotificationType(hlapi.ObjectIdentity(trap_data['oid']))
    var_binds.addVarBinds(
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['alertname']),
                         hlapi.OctetString(trap_data['alertname'] or '')),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['status']),
                         hlapi.OctetString(trap_data['status'] or '')),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['severity']),
                         hlapi.OctetString(trap_data['severity'] or '')),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['instance']),
                         hlapi.OctetString(trap_data['instance'] or '')),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['job']),
                         hlapi.OctetString(trap_data['job'] or '')),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['description']),
                         hlapi.OctetString(trap_data['description'] or '')),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['labels']),
                         hlapi.OctetString(json.dumps(trap_data['labels'] or ''))),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['timestamp']),
                         hlapi.TimeTicks(trap_data['timestamp'])),
        hlapi.ObjectType(hlapi.ObjectIdentity(oids['rawdata']),
                         hlapi.OctetString(json.dumps(trap_data['rawdata'])))
    )

    error_indication, error_status, error_index, _ = next(
        hlapi.sendNotification(
            hlapi.SnmpEngine(),
            hlapi.CommunityData(config['snmp_community'], mpModel=1),
            transport_target, hlapi.ContextData(), 'trap', var_binds))

    if error_indication:
        logger.error('SNMP trap not sent: %s', error_indication)
    elif error_status:
        logger.error('SNMP trap receiver returned error: %s @%s',
                     error_status, error_index)
    else:
        logger.debug('Sending SNMP trap: %s', trap_data)


def get_http_server_config():
    """
    Get the CherryPy application config.
    :return: The CherryPy application config.
    :rtype: dict
    """
    return {'/': {'request.dispatch': cherrypy.dispatch.MethodDispatcher()}}


def run_http_server(ctx):
    """
    Run the HTTP server.
    :param ctx: The application context.
    :type ctx: dict
    """
    if ctx.config['cert'] and ctx.config['key']:
        cherrypy.config.update({
            'server.ssl_module': 'builtin',
            'server.ssl_certificate': ctx.config['cert'],
            'server.ssl_private_key': ctx.config['key']
        })
    elif ctx.config['cert'] or ctx.config['key']:
        raise Exception("Both '--cert' and '--key' are needed")

    cherrypy.config.update({
        'environment': 'production',
        'server.socket_host': ctx.config['host'],
        'server.socket_port': ctx.config['port']
    })
    cherrypy.quickstart(Root(ctx), config=get_http_server_config())


class Config(dict):
    def __init__(self):
        super().__init__()
        self.reset()

    @staticmethod
    def defaults():
        """
        Get the default configuration values.
        :return: Returns a dictionary containing the default values.
        :rtype: dict
        """
        return {
            'debug': False,
            'snmp_host': 'localhost',
            'snmp_port': 162,
            'snmp_community': 'public',
            'snmp_retries': 5,
            'snmp_timeout': 1,
            'alert_oid_label': 'oid',
            'trap_oid_prefix': '1.3.6.1.4.1.50495.15',
            'trap_default_oid': '1.3.6.1.4.1.50495.15.1.2.1',
            'trap_default_severity': '',
            'host': '0.0.0.0',
            'port': 9099,
            'metrics': False,
            'cert': '',
            'key': ''
        }

    def dump(self):
        logger.debug('Configuration settings: %s', json.dumps(self))

    def reset(self, name=None):
        """
        Reset to default values. If a name is specified, only the named
        configuration setting is reset to default.
        :param name: The name of the configuration setting. Defaults to 'None'.
        :type name: str
        """
        if name is None:
            self.clear()
            self.update(Config.defaults())
        else:
            self[name] = Config.defaults()[name]

    def load(self, prog_name):
        """
        Load a configuration file from disk.
        :param prog_name: The name of the program.
        :type prog_name: str
        """
        file_name = '{}.conf'.format(prog_name)
        path_names = [
            os.path.join('/etc', file_name),
            os.path.join(sys.path[0], file_name)
        ]
        for path_name in path_names:
            try:
                with open(path_name, 'r', encoding='UTF-8') as stream:
                    config = yaml.safe_load(stream)
                    # Automatically convert hyphens to underscores.
                    for key in list(config.keys()):
                        config[key.replace('-', '_')] = config.pop(key)
                    self.update(config)
            except (IOError, FileNotFoundError):
                pass

    def __setitem__(self, key, value):
        """
        Set self[key] to value. Ignore 'None' values.
        :param key: The name of the key.
        :type key: str
        :param value: The value of the key.
        :type value: bool|int|str
        """
        if value is not None:
            super().__setitem__(key, value)


class Telemetry:
    def __init__(self):
        self.metrics = {
            'notifications': prometheus_client.Counter(
                'prometheus_webhook_snmp_notifications',
                'Number of processed Prometheus Alertmanager notifications.'),
            'traps': prometheus_client.Counter(
                'prometheus_webhook_snmp_traps',
                'Number of sent SNMP traps.')
        }

    def inc(self, name):
        self.metrics[name].inc()

    def generate(self):
        result = []
        for _, metric in self.metrics.items():
            result.append(prometheus_client.generate_latest(metric).
                          decode('utf-8'))
        return ''.join(result)


class Context:
    def __init__(self):
        self.config = Config()
        self.telemetry = Telemetry()


class Metrics:  # pylint: disable=too-few-public-methods
    exposed = True

    def __init__(self, ctx):
        """
        :param ctx: The application context.
        :type ctx: dict
        """
        self.ctx = ctx

    def GET(self):
        """
        Get Prometheus metrics from this receiver.
        :return: The Prometheus metrics.
        :rtype: str
        """
        if not self.ctx.config['metrics']:
            raise cherrypy.HTTPError(405)

        cherrypy.response.headers['Content-Type'] = 'text/plain'
        return self.ctx.telemetry.generate()


class Root:  # pylint: disable=too-few-public-methods
    exposed = True

    def __init__(self, ctx):
        """
        :param ctx: The application context.
        :type ctx: dict
        """
        self.ctx = ctx
        self.metrics = Metrics(ctx)

    @cherrypy.tools.json_in()
    def POST(self):
        """
        Listen for Prometheus Alertmanager notifications.
        """
        notification = cherrypy.request.json
        logger.debug('Receiving notification: %s', notification)
        self.ctx.telemetry.inc('notifications')
        trap_data = parse_notification(self.ctx.config, notification)
        if trap_data is not None:
            for data in trap_data:
                send_snmp_trap(self.ctx.config, data)
                self.ctx.telemetry.inc('traps')
